Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,149 @@
+{
+    m_masonryAxisDirection = masonryAxisDirection;
+    m_gridAxisDirection = m_masonryAxisDirection == ForRows ? ForColumns : ForRows;
+    m_masonryAxisGridGap = m_renderGrid.gridGap(m_masonryAxisDirection);
+    m_gridAxisTracksCount = m_renderGrid.numTracks(m_gridAxisDirection);
+    m_runningPositions = Vector<LayoutUnit>(m_gridAxisTracksCount);
+
+    // 2.4 Masonry Layout Algorithm
+    addItemsToFirstTrack(firstTrackItems);
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto& itemItr : firstTrackItems) {
+        auto* item = itemItr.key;
+        auto gridArea = itemItr.value;
+
+        insertIntoGridAndLayoutItem(item, gridArea);
+
+        updateRunningPositions(item);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(Vector<RenderBox*> itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* item : itemsWithDefiniteGridAxisPosition) {
+        auto itemSpan = m_renderGrid.currentGrid().gridItemSpan(*item, m_gridAxisDirection);
+
+        ASSERT(!itemSpan.isIndefinite());
+
+        auto initialTrackSpan = GridSpan::translatedDefiniteGridSpan(0, 1);
+        itemSpan.translate(m_renderGrid.currentGrid().explicitGridStart(m_gridAxisDirection));
+
+        GridArea area = m_masonryAxisDirection == ForRows ? GridArea(initialTrackSpan, itemSpan) : GridArea(itemSpan, initialTrackSpan);
+        insertIntoGridAndLayoutItem(item, area);
+
+        updateRunningPositions(item);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition(Vector<RenderBox*> itemsWithIndefinitePosition)
+{
+    for (auto* item : itemsWithIndefinitePosition) {
+        insertIntoGridAndLayoutItem(item, nextMasonryPositionForItem(item));
+
+        updateRunningPositions(item);
+    }
+}
+
+void GridMasonryLayout::insertIntoGridAndLayoutItem(RenderBox* child, GridArea area)
+{
+    m_renderGrid.currentGrid().insert(*child, area);
+
+    if (m_gridAxisDirection == ForColumns)
+        m_renderGrid.updateGridAreaLogicalSize(*child, m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns), std::nullopt);
+    else
+        m_renderGrid.updateGridAreaLogicalSize(*child, std::nullopt, m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
+    child->layoutIfNeeded();
+}
+
+void GridMasonryLayout::updateRunningPositions(RenderBox* child)
+{
+    auto gridSpan = m_masonryAxisDirection == ForRows ? m_renderGrid.currentGrid().gridItemArea(*child).columns : m_renderGrid.currentGrid().gridItemArea(*child).rows;
```

Comment:
Probably it would be more readable like this:
```cpp
GridArea area = currentGrid().gridItemArea(*child);
auto gridSpan = m_masonryAxisDirection == ForRows ? area.columns : area.rows;
```

Or, the callers already have the GridArea, so it could also be passed as a parameter like in `insertIntoGridAndLayoutItem`.

In fact, `insertIntoGridAndLayoutItem` and `updateRunningPositions` might even be merged? But keep them separate if you think it makes more sense like this.

Reply:
Yes, I think it makes sense to merge `updateRunningPositions` into `insertIntoGridAndLayoutItem` since we should only be updating the running positions whenever we insert into the grid for placement. I kept the method, however, since I think it still makes sense to keep the logic contained within it.